pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -150,25 +192,58 @@ func (state PipelineRunState) ToMap() map[string]*ResolvedPipelineRunTask {
 	return m
 }
 
-// IsDone returns true when all pipeline tasks have respective taskRun created and
-// that taskRun has either succeeded or failed after all possible retry attempts
-func (state PipelineRunState) IsDone() (isDone bool) {
-	isDone = true
+// HsFailedOrCancelled returns true the state includes at least one failed or one cancelled task
+func (state PipelineRunState) HasFailedOrCancelled() bool {
 	for _, t := range state {
-		if t.TaskRun == nil || t.PipelineTask == nil {
+		if t.IsCancelled() {
+			return true
+		}
+		if t.IsFailure() {
+			return true
+		}
+	}
+	return false
+}
+
+// State returned the combined state for the PipelineRun
+func (rpr ResolvedPipelineRun) State() PipelineRunState {
+	return append(rpr.DagState, rpr.FinalState...)
+}
+
+// isDAGDone returns true if no task from the DAG is running and no new one can be scheduled
+func (rpr ResolvedPipelineRun) isDAGDone() bool {
+	for _, t := range rpr.DagState {
+		if t.TaskRun == nil {
+			if rpr.IsTaskSkipped(t.PipelineTask.Name) {
+				continue
+			}
 			return false
 		}
-		isDone = isDone && t.IsDone()
-		if !isDone {
-			return
+		if !t.IsDone() {
+			return false
 		}
 	}
-	return
+	return true
+}
+
+// isFinallyDone returns true if all final tasks are done
+func (rpr ResolvedPipelineRun) isFinallyDone() bool {
```

Comment:
This is only used as part of `IsDone` below, so I could move it in there... I thought it was more readable like this though.